docs/source/development/eep-01-pytrees.rst

```diff
@@ -0,0 +1,848 @@
+    }
+
+
+Run an optimization
+-------------------
+
+.. code-block:: python
+
+    from estimagic import minimize
+
+    minimize(
+        criterion=crit,
+        params=params,
+        algorithm="scipy_lbfgsb",
+    )
+
+The internal optimizer (in this case the lbfgsb algorithm from scipy) will see a wrapped
+version of ``crit``. That version takes a 1d numpy array as its only argument and
+returns a scalar float (the ``"value"`` entry of the result of ``crit``). Numerical
+derivatives are also taken on that function.
+
+If instead a derivative based least squares optimizer like ``"scipy_ls_dogbox"`` had
+been used, the internal optimizer would see a modified version of ``crit`` that takes
+a 1d numpy array and returns a 1d numpy array (the flattened version of the
+``"root_contributions"`` entry of the result of ``crit``).
+
+
+The optimization output
+-----------------------
+
+The following entries of the output of minimize are affected by the change:
+
+- ``"solution_params"``: A pytree with the same structure as ``params``
+- ``"solution_criterion"``: The output dictionary of ``crit`` evaluated solution params
+- ``solution_derivative``: Maybe we should not even have this entry.
+
+.. danger:: We need to discuss if an in which form we want to have a solution
+    derivative entry. In it's current form it is useless if constraints are used.
+    This gets worse when we allow for pytrees and translating this into a meaningful
+    shape might be very difficult.
+
+
+Add bounds
+----------
+
+Bounds on parameters that are inside a DataFrame with ``"value"`` column can simply be
+specified as before. For all others, there are separate ``lower_bounds`` and
+``upper_bounds`` arguments in ``maximize`` and ``minimize``.
+
+``lower_bounds`` and ``upper_bounds`` are pytrees of the same structure as ``params`` or
+a subtree that preserves enough structure to match all bounds. For example:
+
+.. code-block:: python
+
+    minimize(
+        criterion=crit,
+        params=params,
+        algorithm="scipy_lbfgsb",
+        lower_bounds={"delta": 0},
+        upper_bounds={"delta": 1},
+    )
+
+This would add bounds for delta, keep the bounds on all ``"utility"`` parameters, and
+leave the ``"probs"`` parameters unbounded.
+
+
+Add a constraint
+----------------
+
+Currently, parameters to which a constraint is applied are selected via a ``"loc"`` or
+``"query"`` entry in the constraints dictionary.
+
+This keeps working as long as params are specified as a single DataFrame containing a
+``"value"`` column. If a more general pytree is used we need a "selector" entry instead.
+The value of that entry is a callable that takes the pytree and returns selected
+parameters.
+
+The selected parameters can be returned as estimagic-pytrees. For constraints where
+order plays a role (e.g., increasing), the order defined by ``tree_flatten`` is used.
+However we advise users to return one-dimensional arrays in that case, so they do not
```

Comment:
The return of which function precisely is meant?

Reply:
Return of the selector function. It's clarified now.